crates/red_knot_ide/src/lib.rs

```diff
@@ -1,3 +1,222 @@
+    fn navigation_targets(&self, db: &dyn Db) -> NavigationTargets;
+}
+
+impl HasNavigationTargets for Type<'_> {
+    fn navigation_targets(&self, db: &dyn Db) -> NavigationTargets {
+        match self {
+            Type::FunctionLiteral(function) => function.navigation_targets(db),
+            Type::ModuleLiteral(module) => module.navigation_targets(db),
+            Type::Union(union) => union
+                .iter(db.upcast())
+                .flat_map(|target| target.navigation_targets(db))
+                .collect(),
+            Type::ClassLiteral(class) => class.navigation_targets(db),
+            Type::Instance(instance) => instance.navigation_targets(db),
+            Type::KnownInstance(instance) => instance.navigation_targets(db),
+            Type::StringLiteral(_)
+            | Type::AlwaysTruthy
+            | Type::AlwaysFalsy
+            | Type::IntLiteral(_)
+            | Type::BooleanLiteral(_)
+            | Type::LiteralString
+            | Type::BytesLiteral(_)
+            | Type::SliceLiteral(_) => self.to_meta_type(db.upcast()).navigation_targets(db),
+
+            Type::Dynamic(_)
+            | Type::SubclassOf(_)
+            | Type::Never
+            | Type::Callable(_)
+            | Type::Intersection(_)
+            | Type::Tuple(_) => NavigationTargets::empty(),
+        }
+    }
+}
+
+impl HasNavigationTargets for FunctionType<'_> {
+    fn navigation_targets(&self, db: &dyn Db) -> NavigationTargets {
+        let function_range = self.focus_range(db.upcast());
+        NavigationTargets::single(NavigationTarget {
+            file: function_range.file(),
+            focus_range: function_range.range(),
+            full_range: self.full_range(db.upcast()).range(),
+        })
+    }
+}
+
+impl HasNavigationTargets for ClassLiteralType<'_> {
+    fn navigation_targets(&self, db: &dyn Db) -> NavigationTargets {
+        let class = self.class();
+        let class_range = class.focus_range(db.upcast());
+        NavigationTargets::single(NavigationTarget {
+            file: class_range.file(),
+            focus_range: class_range.range(),
+            full_range: class.full_range(db.upcast()).range(),
+        })
+    }
+}
+
+impl HasNavigationTargets for InstanceType<'_> {
+    fn navigation_targets(&self, db: &dyn Db) -> NavigationTargets {
+        let class = self.class();
+        let class_range = class.focus_range(db.upcast());
+        NavigationTargets::single(NavigationTarget {
+            file: class_range.file(),
+            focus_range: class_range.range(),
+            full_range: class.full_range(db.upcast()).range(),
+        })
+    }
+}
+
+impl HasNavigationTargets for ModuleLiteralType<'_> {
+    fn navigation_targets(&self, db: &dyn Db) -> NavigationTargets {
+        let file = self.module(db).file();
+
+        NavigationTargets::single(NavigationTarget {
+            file,
+            focus_range: TextRange::default(),
+            full_range: TextRange::default(),
```

Comment:
Should this be the entire module range?

Reply:
Possibly. It makes it a bit more awkward to implement because I then need to query the source text (it's not hard). But it's not clear to me if that's more accurate because the module doesn't strictly have a range. It's just the file. That's why I opted for the most trivial solution for now. We can still adjust the ranges if it turns out that we should highlight the entire module